ksql-engine/src/main/java/io/confluent/ksql/engine/TopicAccessValidator.java

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright 2019 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.engine;
+
+import avro.shaded.com.google.common.collect.Sets;
+import io.confluent.ksql.parser.tree.CreateAsSelect;
+import io.confluent.ksql.parser.tree.CreateSource;
+import io.confluent.ksql.parser.tree.InsertInto;
+import io.confluent.ksql.parser.tree.InsertValues;
+import io.confluent.ksql.parser.tree.Statement;
+import io.confluent.ksql.planner.plan.DataSourceNode;
+import io.confluent.ksql.planner.plan.PlanNode;
+import io.confluent.ksql.services.ServiceContext;
+import io.confluent.ksql.util.KsqlTopicAccessException;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+
+import org.apache.kafka.clients.admin.TopicDescription;
+import org.apache.kafka.common.acl.AclOperation;
+
+/**
+ * Wraps a {@link ServiceContext} to validate access to Kafka topics.
+ */
+public final class TopicAccessValidator {
+  private final ServiceContext serviceContext;
+
+  private TopicAccessValidator(final ServiceContext serviceContext) {
+    this.serviceContext = serviceContext;
+  }
+
+  /**
+   * Creates a {@code TopicAccessValidator} that uses the specified {@code ServiceContext}
+   * to check Kafka topic access.
+   *
+   * @param serviceContext The {@code ServiceContext} to use.
+   * @return A {@code TopicAccessValidator} object.
+   */
+  public static TopicAccessValidator as(final ServiceContext serviceContext) {
+    return new TopicAccessValidator(serviceContext);
+  }
+
+  /**
+   * Verifies if all query source topics have read access.
+   *
+   * @param node The node generated by the {@code QueryEngine#buildLogicalPlan}.
+   * @throws KsqlTopicAccessException If a topic does not exist or does not have permissions.
+   */
+  public void verifyQuerySourcesPermissions(final PlanNode node) throws KsqlTopicAccessException {
+    final Set<String> kafkaTopics = new HashSet<>();
+    collectKafkaTopics(kafkaTopics, node);
+
+    for (String kafkaTopic : kafkaTopics) {
+      checkAccess(kafkaTopic, Collections.singleton(AclOperation.READ));
+    }
+  }
+
+  /**
+   * Verifies if a non-query KSQL statement has permissions to access the required topics.
+   *
+   * @param statement The statement to verify for permissions.
+   * @throws KsqlTopicAccessException If a topic does not exist or does not have permissions.
+   */
+  public void verifyStatementPermissions(final Statement statement)
```

Comment:
I would do this slightly differently - I think it makes sense for all of these Statements to implement an interface (e.g. `TopicAccessor`) which has two methods: `getReadTopics()` and `getWriteTopics()`. then this class will accept a `TopicAccessor` and check all permissions for anything that implements that. 

Another benefit of this aside from cleaner code is that we could reuse that for anything (not just Statements - e.g. `PlanNode` can implement this too)

Reply:
That's a nice idea! But now that I reduced the number of statements to 2, I think that will just add more code that needs more testing and review. I might do it in a follow-up PR.